pkg/config/lifecycle_provider.go

```diff
@@ -0,0 +1,124 @@
+package config
+
+import (
+	"sync/atomic"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	LifecycleConfigName = "lifecycle-image"
+	LifecycleConfigKey  = "image"
+)
+
+type RegistryClient interface {
+	Fetch(keychain authn.Keychain, repoName string) (v1.Image, string, error)
+}
+
+type lifecycleData struct {
+	image v1.Image
+	err   error
+}
+
+type LifecycleProvider struct {
+	RegistryClient RegistryClient
+	Keychain       authn.Keychain
+	lifecycleData  atomic.Value
+	handlers       []func()
+}
+
+func NewLifecycleProvider(lifecycleImageRef string, client RegistryClient, keychain authn.Keychain) *LifecycleProvider {
+	p := &LifecycleProvider{
+		RegistryClient: client,
+		Keychain:       keychain,
+	}
+
+	data := &lifecycleData{}
+
+	data.image, data.err = p.fetchImage(lifecycleImageRef)
+
+	p.lifecycleData.Store(data)
+	return p
+}
+
+func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
+	data := &lifecycleData{}
+	isNewImg := true
+	defer l.callCallBacks(&isNewImg)
+	defer l.lifecycleData.Store(data)
+
+	imageRef, ok := cm.Data[LifecycleConfigKey]
+	if !ok {
+		data.err = errors.Errorf("%s config invalid", LifecycleConfigName)
+		return
+	}
+
+	newImg, err := l.fetchImage(imageRef)
```

Comment:
A good amount of the complexity in this func is due it not returning an err. Could we avoid that complexity by calling a method that does return an error and only error handle here?
```
func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
    defer l.callCallBacks(&isNewImg)
    data := &lifecycleData{}
    data.img, data.err = readLifecycleImage(cm)
    l.lifecycleData.Store(data)
}
```

Reply:
Perhaps we could take a page from reconcilers:
```
func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
    data := &lifecycleData{}
    data.img, data.err = readLifecycleImage(cm)
    if !equality.DeepEqual(l.lifecycleData.get(), data) {
        l.lifecycleData.Store(data)
        l.callCallbacks()

   }
}
```